### ISSUE TITLE
When I set the module to node16, why do I need to import .js instead of .ts files? At this time, js has not been generated yet.

### ISSUE THREAD

**Issue by alamhubb** [microsoft/TypeScript#60319]
### 🔍 Search Terms

When I set the module to node16, why do I need to import .js instead of .ts files? At this time, js has not been generated yet.

I feel that the logic of typescript is very strange. Why do I need to set the suffix of the imported file to the uncompiled .js file instead of .ts during development? Why not import the *.ts file suffix during development, and then automatically convert it to .js after compilation? I feel that this logic is very self-consistent. Why doesn't typescript adopt this logic?

https://github.com/microsoft/TypeScript/issues/60277

https://github.com/microsoft/TypeScript/issues/40878

https://github.com/microsoft/TypeScript/issues/41887

### ✅ Viability Checklist

- [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
- [x] This wouldn't change the runtime behavior of existing JavaScript code
- [x] This could be implemented without emitting different JS based on the types of the expressions
- [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
- [x] This isn't a request to add a new utility type: https://github.com/microsoft/TypeScript/wiki/No-New-Utility-Types
- [x] This feature would agree with the rest of our Design Goals: https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals

### ⭐ Suggestion

When I set the module to node16, why do I need to import .js instead of .ts files? At this time, js has not been generated yet.

### 📃 Motivating Example

When I set the module to node16, why do I need to import .js instead of .ts files? At this time, js has not been generated yet.

### 💻 Use Cases

When I set the module to node16, why do I need to import .js instead of .ts files? At this time, js has not been generated yet.

**Comment by jakebailey** [comment]
Consider reading https://www.typescriptlang.org/docs/handbook/modules/theory.html

**Comment by jakebailey** [comment]
I'm not sure why you're opening the same issue twice; you're just going to get the same answer...

**Comment by alamhubb** [comment]
> I'm not sure why you're opening the same issue twice; you're just going to get the same answer...

Thanks for your answers

The first time I encountered this problem, I got the answer, but I was confused about the answer, but there was no follow-up. This time I directly expressed my doubts about the answer.

I expect that .ts files can be directly imported into ts and converted to .js after compilation, instead of directly importing .js. I can't understand the reason for doing this and find it strange. I think it is more logical to introduce .ts during development and convert it to .js after compilation.

**Comment by alamhubb** [comment]
> Consider reading https://www.typescriptlang.org/docs/handbook/modules/theory.html

Importing ungenerated .js in ts is counter-intuitive and even illogical. No matter how unfathomable the reason behind it is, I think it is unreasonable.

I read the document about cjs, cts, mjs, and mts. I can't understand why it can be compatible without extension symbols, but ts can't do it with extension symbols. I believe it can be done.

If the extension symbol ts can be used uniformly, the code before and after compilation can be consistent without modification. I think this is more convenient.

**Comment by jakebailey** [comment]
> I think it is more logical to introduce .ts during development and convert it to .js after compilation.

TypeScript describes executable code. Code with .ts imports are not executable because the .ts files will not exist after TypeScript emits .js files.

But, see also https://devblogs.microsoft.com/typescript/announcing-typescript-5-7-beta/#path-rewriting-for-relative-paths

**Comment by alamhubb** [comment]
> TypeScript describes executable code. Code with .ts imports are not executable because the .ts files will not exist after TypeScript emits .js files.
> 
> But, see also https://devblogs.microsoft.com/typescript/announcing-typescript-5-7-beta/#path-rewriting-for-relative-paths

Thank you very much for answering my question. I thought that since I am a beginner and I don't have a deep understanding of ts, you may not want to answer. So thank you very much.

Due to your answer, I also want to say more about my inner thoughts. It may not be very friendly. I hope you can forgive me.

I carefully checked the link you sent me, but my English is not very good, but I think I roughly understood it. The relevant problem should be in this part. ts will not rewrite the path.

> There is some complexity to be aware of when using these modes though. To be maximally compatible with all these tools, a TypeScript file that’s imported "in-place" must be imported with the appropriate TypeScript extension at runtime. For example, to import a file called foo.ts, we have to write the following in Node’s new experimental support:

> Copy
// main.ts

> import * as foo from "./foo.ts"; // <- we need foo.ts here, not foo.js
Typically, TypeScript would issue an error if we did this, because it expects us to import the output file. Because some tools do allow .ts imports, TypeScript has supported this import style with an option called --allowImportingTsExtensions for a while now. This works fine, but what happens if we need to actually generate .js files out of these .ts files? This is a requirement for library authors who will need to be able to distribute just .js files, but up until now TypeScript has avoided rewriting any paths.

Then I still want to emphasize my point again. For a beginner user, this problem is a serious bug in the language and a serious program vulnerability, which can directly discourage beginners.

Or let us understand this problem in this way. Don't bind ts and js. ts is just ts, a new programming language, and it has nothing to do with js. If we look at this problem from this perspective, will we still think that these difficulties are justified and do not need to be overcome?

I think the content of the article does not treat the position of ts correctly. It has always viewed this matter as a binding relationship with js, so many problems can be attributed to the association with js

I am using ts, and there are problems when I introduce a ts file. This is a big deal

Of course, we may have different perceptions of the positioning of ts. I think ts should be able to run as a language unrelated to js (at least it looks and should be used like this, so that we can completely stay away from the annoying js). I hope ts can look and run unrelated to js. I think it will become a better ts. If there is any problem, it is because of the association with js. I think its development and design will be restricted, but I also support ts's friendly design that is fully compatible with js. This is good, but it does not make it look related to js. It can only be related to syntax and support the import of .js files. I hope ts can solve these problems in a better way

Of course, I am not ts As the maintainer of , I don't work hard, I just talk here, I just put forward my suggestions as a beginner, hoping to be adopted

After picking so many questions, I sincerely thank all ts developers and senior programming engineers for their hard work. Thank you, we enjoy the fruits of your work, respect

**Comment by alamhubb** [comment]
Related questions

https://github.com/microsoft/TypeScript/issues/60288#issuecomment-2428001136

**Comment by MartinJohns** [comment]
> ts is just ts, a new programming language, and it has nothing to do with js.

That is exactly what it is **not**. See the [TypeScript Design Goals](https://github.com/microsoft/TypeScript/wiki/TypeScript-Design-Goals).

Or just the headline from the [website](https://www.typescriptlang.org/):

> TypeScript is JavaScript with syntax for types.

**Comment by alamhubb** [comment]
> > ts is just ts, a new programming language, and it has nothing to do with js.
> 
> That is exactly what it is **not**. See the [TypeScript Design Goals](https://github.com/microsoft/TypeScript/wiki/TypeScript-Design-Goals).
> 
> Or just the headline from the [website](https://www.typescriptlang.org/):
> 
> > TypeScript is JavaScript with syntax for types.

What I mean is that ts should give people the feeling that it looks like a brand new language, so that users don't need to care about js

It doesn't mean that its implementation principle has nothing to do with js

What its underlying structure is and what it looks like after compilation should not be of concern to users. I feel that this is the ideal appearance of ts

**Comment by TsXor** [comment]
Actually, anyone should completely ignore this issue, as @alamhubb is actually using issue section as after-sales service.
In his opinion, the TypeScript language is just a product free of charge. Now that it is a product, its developers are in the role of producers and he is in the role of customer, and you should serve him well because “Customer is God”.
That is his logic. Your product is so hard to use! Of course you should immediately fix that whatever the cost! Don't say nonsense like “This is not TypeScript's goal”! JUST FIX IT!
Clearly, the root of such logic it not even wrong.

How did I know this is his logic? See [this question on ZhiHu](https://www.zhihu.com/question/1853777085). He just cannot understand why you developers can annoy him so much that he went to ask others if he did it right.

**Comment by alamhubb** [comment]
> Actually, anyone should completely ignore this issue, as [@alamhubb](https://github.com/alamhubb) is actually using issue section as after-sales service. In his opinion, the TypeScript language is just a product free of charge. Now that it is a product, its developers are in the role of producers and he is in the role of customer, and you should serve him well because “Customer is God”. That is his logic. Your product is so hard to use! Of course you should immediately fix that whatever the cost! Don't say nonsense like “This is not TypeScript's goal”! JUST FIX IT! Clearly, the root of such logic it not even wrong.
> 
> How did I know this is his logic? See [this question on ZhiHu](https://www.zhihu.com/question/1853777085). He just cannot understand why you developers can annoy him so much that he went to ask others if he did it right.

js also has many problems, which are clearly written in the book, but users are still often troubled by these problems. So is this a problem that developers did not do enough homework? I understand that a good product should allow users to walk normally and not hit the south wall. I think my demand is very normal, that is, to introduce ts during ts development and compile it into js. I think this is a very normal way, so I put forward my demand. I don’t think that making reasonable demands is embarrassing for maintainers. Of course, the final decision is not up to me. I just express my opinion. Whether it is accepted or not is not up to me.

**Comment by TsXor** [comment]
> js also has many problems, which are clearly written in the book, but users are still often troubled by these problems. So is this a problem that developers did not do enough homework? I understand that a good product should allow users to walk normally and not hit the south wall. I think my demand is very normal, that is, to introduce ts during ts development and compile it into js. I think this is a very normal way, so I put forward my demand. I don’t think that making reasonable demands is embarrassing for maintainers. Of course, the final decision is not up to me. I just express my opinion. Whether it is accepted or not is not up to me.

Open source projects are NOT products. No one owes you usability and convenience. If you want them, use your hands to make them.
You should not “make demands”. You should open feature requests to discuss if you demand is common and necessary to be solved, or try to implement what you want and open a pull request.

**Comment by alamhubb** [comment]
> > js also has many problems, which are clearly written in the book, but users are still often troubled by these problems. So is this a problem that developers did not do enough homework? I understand that a good product should allow users to walk normally and not hit the south wall. I think my demand is very normal, that is, to introduce ts during ts development and compile it into js. I think this is a very normal way, so I put forward my demand. I don’t think that making reasonable demands is embarrassing for maintainers. Of course, the final decision is not up to me. I just express my opinion. Whether it is accepted or not is not up to me.
> 
> Open source projects are NOT products. No one owes you usability and convenience. If you want them, use your hands to make them. You should not “make demands”. You should open feature requests to discuss if you demand is common and necessary to be solved, or try to implement what you want and open a pull request.

I am making a request, issues supports this function, you can carefully look at the support list of issues when you create,

The point is whether this is a reasonable need, I think it is, if not why 5.7 should support it

The origin of the problem, because I am the first and second questions, didn't get 5.7 rewriteRelativeImportExtensions answers, but has been unable to solve

I think this is a very reasonable demand, ts should support him, actually 5.7 he did, but did not say, but kept saying some reasons can not be supported

Then I hope I can look at the problem from a different perspective and see if the problem is really unsolvable or evasive

Can you make sure that all the problems you encounter with open source projects, you don't mention issues, but your own pr? Human energy is limited, I am also maintaining my own open source library, but also in my own way to contribute, everyone focus on their own field, I think it is good

**Comment by alamhubb** [comment]
This is a reasonable requirement and is already supported in version 5.7. The reason for not supporting it has expired. Just enable rewriteRelativeImportExtensions in version 5.7.

**Comment by YangRucheng** [comment]
> > 考虑阅读 https://www.typescriptlang.org/docs/handbook/modules/theory.html
> 
> 在 ts 中导入未生成的 .js 是违反直觉的，甚至是不合逻辑的。无论背后的原因多么深不可测，我认为都是不合理的。
> 
> 我阅读了有关 cjs、cts、mjs 和 mts 的文档。我不明白为什么它可以在没有扩展符号的情况下兼容，但 ts 不能用扩展符号来做到这一点。我相信这是可以做到的。
> 
> 如果扩展符号 ts 可以统一使用，则编译前后的代码可以保持一致，无需修改。我觉得这样更方便。



> > 考虑阅读 https://www.typescriptlang.org/docs/handbook/modules/theory.html
> 
> 在 ts 中导入未生成的 .js 是违反直觉的，甚至是不合逻辑的。无论背后的原因多么深不可测，我认为都是不合理的。
> 
> 我阅读了有关 cjs、cts、mjs 和 mts 的文档。我不明白为什么它可以在没有扩展符号的情况下兼容，但 ts 不能用扩展符号来做到这一点。我相信这是可以做到的。
> 
> 如果扩展符号 ts 可以统一使用，则编译前后的代码可以保持一致，无需修改。我觉得这样更方便。




> > 我不确定你为什么两次打开同一个问题;你只会得到同样的答案......
> 
> 感谢您的回答
> 
> 第一次遇到这个问题，我得到了答案，但是我对答案感到困惑，但是没有后续。这一次我直接表达了我对答案的疑惑。
> 
> 我希望.ts文件可以直接导入到 ts 中，编译后转换成 .js，而不是直接导入 .js。我不明白这样做的原因，觉得很奇怪。我认为在开发过程中引入 .ts 并在编译后将其转换为 .js 更合乎逻辑。

t

> > Consider reading https://www.typescriptlang.org/docs/handbook/modules/theory.html
> 
> Importing ungenerated .js in ts is counter-intuitive and even illogical. No matter how unfathomable the reason behind it is, I think it is unreasonable.
> 
> I read the document about cjs, cts, mjs, and mts. I can't understand why it can be compatible without extension symbols, but ts can't do it with extension symbols. I believe it can be done.
> 
> If the extension symbol ts can be used uniformly, the code before and after compilation can be consistent without modification. I think this is more convenient.

ts 不能执行，也没有导入行为。导入是在编译成 js 后，在 js 中进行的。

**Comment by YangRucheng** [comment]
> Or, if you offer a reward of 1 million US dollars to solve this problem, you need to support importing other ts files in ts. Can it be solved?
> So why do you think it can't be solved now?

I believe that @alamhubb 's comments violate the Microsoft Code of Conduct.

As the leader of the project, have you considered correcting or deleting his comments to maintain a positive and inspiring community?

Specifically, his remarks may violate the principle of "making participation in our community a harassment-free experience for everyone," as he attempts to leverage a supposed economic advantage to make demands of the community.

Thank you for your patience. @jakebailey

**Comment by alamhubb** [comment]
Based on the discussion about this requirement on Zhihu, the Chinese intranet, I think there are some misunderstandings about the maintainer in this issue. I would like to clarify and apologize again.

After a few days of understanding, I think this issue can be described like this. In fact, I want to propose a requirement, I want ts to support the .ts file imported into the file, and automatically compile it into .js (that is, the function of automatically rewriting the path, but my ability makes me unable to accurately describe this requirement)

I think this is a very important function. TS does not support this function, which is a big blow to beginners. TS has gradually become the infrastructure in the front-end field.

Even though I checked the related issues before and saw that the ts maintainer explicitly rejected this requirement, in 2020, I still thought that this requirement was more important from my perspective, and 4 years have passed, so I raised the issue again.

Due to the subject content and expression of my issues, the TS maintainer was unable to understand my expression correctly (there was a problem with my expression, which was not direct enough)

Since my question asked why, the maintainer gave me the answer about why it is not possible, but what I actually wanted to say was that I hope TS can support this feature, and then the maintainer told me some reasons why it is not possible. But in my opinion, this feature is really important, and it is not a feature that the maintainer cannot implement. After all, the technical level of the maintainers is very excellent in the industry.

The maintainer asked me why I asked the same question twice. I explained it myself and this time directly stated what I expected.

Then I expressed the demand directly this time, and the maintainer understood what I meant, so he posted an article about 5.7 supporting this feature for me, but the maintainer again stated that ts could not be imported into ts, and then I mistakenly thought that the developer sent me another reason why ts could not support this feature, and I found relevant content in the article

Based on my belief that this feature can be implemented, I made some more radical remarks, trying to prove from other levels that this demand is reasonable and achievable, but in fact I did not find that the article already included the content that 5.7 version supports this feature

Because I think a very important demand is not supported by ts yet, and the official team believes that this demand is unreasonable, I went to ask questions in the Zhihu community in China, hoping to get more people's opinions, because my technical level is indeed It is relatively low. I want to know why the official team thinks this requirement is unreasonable. I got some relevant answers in the Zhihu community and posted a link to this issue.

Then most Chinese netizens criticized my inappropriate remarks. The examples of if it is the boss and if there is a 1 million reward are not appropriate. If I change to the perspective of if this requirement cannot be met, the earth will explode next month, whether this requirement can be met, this kind of remarks may appear more friendly and humorous.

Then after a few days of discussion with Chinese netizens, I learned more about the content of this requirement.

If any of my remarks in this issue caused any discomfort to the maintainers of ts, I express my deep apologies. The misunderstanding in this issue is caused by my personal incompetence. I am very sorry.
